core/evtforward/ethcall/engine.go

```diff
@@ -28,71 +25,229 @@ type Forwarder interface {
-	return &Engine{
+func NewEngine(log *logging.Logger, cfg Config, client EthReaderCaller, forwarder Forwarder) *Engine {
+	e := &Engine{
 		log:       log,
 		cfg:       cfg,
 		client:    client,
 		forwarder: forwarder,
-		specs:     make(map[string]Spec),
-	}, nil
+		calls:     make(map[string]Call),
+		poller:    newPoller(cfg.PollEvery.Get()),
+	}
+
+	return e
 }
 
-func (e *Engine) AddSpec(s Spec) (string, error) {
-	id := s.HashHex()
-	e.specs[id] = s
-	return id, nil
+// Start starts the polling of the Ethereum bridges, listens to the events
+// they emit and forward it to the network.
+func (e *Engine) Start() {
+	ctx, cancelEthereumQueries := context.WithCancel(context.Background())
+	defer cancelEthereumQueries()
+
+	e.cancelEthereumQueries = cancelEthereumQueries
+
+	if e.log.IsDebug() {
+		e.log.Debug("Starting ethereum contract call polling engine")
+	}
+
+	e.poller.Loop(func() {
+		e.Poll(ctx, time.Now())
+	})
+}
+
+func (e *Engine) UpdatePreviousEthBlock(height uint64, time uint64) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	e.prevEthBlock = blockIndex{number: height, time: time}
 }
 
-func (e *Engine) GetSpec(id string) (Spec, bool) {
-	spec, ok := e.specs[id]
-	return spec, ok
+func (e *Engine) initPreviousEthBlock(height uint64, time uint64) bool {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.prevEthBlock == nil {
+		e.prevEthBlock = blockIndex{number: height, time: time}
+		return true
+	}
+	return false
+}
+
+func (e *Engine) getPreviousEthBlock() blockish {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	return blockIndex{
+		number: e.prevEthBlock.NumberU64(),
+		time:   e.prevEthBlock.Time(),
+	}
 }
 
-func (e *Engine) OnTick(ctx context.Context, vegaTime time.Time) {
-	block, err := e.client.BlockByNumber(ctx, nil)
+func (e *Engine) getCalls() map[string]Call {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	calls := map[string]Call{}
+	for specID, call := range e.calls {
+		calls[specID] = call
+	}
+	return calls
+}
+
+func (e *Engine) Stop() {
+	// Notify to stop on next iteration.
+	e.poller.Stop()
+	// Cancel any ongoing queries against Ethereum.
+	if e.cancelEthereumQueries != nil {
+		e.cancelEthereumQueries()
+	}
+}
+
+func (e *Engine) GetSpec(id string) (types.EthCallSpec, bool) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if source, ok := e.calls[id]; ok {
+		return source.spec, true
+	}
+
+	return types.EthCallSpec{}, false
+}
+
+func (e *Engine) MakeResult(specID string, bytes []byte) (Result, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	call, ok := e.calls[specID]
+	if !ok {
+		return Result{}, fmt.Errorf("no such specification: %v", specID)
+	}
+	return newResult(call, bytes)
+}
+
+func (e *Engine) CallSpec(ctx context.Context, id string, atBlock uint64) (Result, error) {
+	e.mu.Lock()
+	call, ok := e.calls[id]
+	if !ok {
+		e.mu.Unlock()
+		return Result{}, fmt.Errorf("no such specification: %v", id)
+	}
+	e.mu.Unlock()
+
+	// TODO: check if it's safe to share an eth client between threads
```

Comment:
Minor: I'm guessing it is?

Reply:
(after some digging, we are confident it is indeed threadsafe; that handler is used internally go the client library in a single goroutine and it pushes stuff to it on a channel)